File: haskell/glue/src/Glue/Lib/List/Foldl.hs
module Glue.Lib.List.Foldl where

import Control.Monad (foldM)
import Glue.Eval (Eval, apply, throwError)
import Glue.Eval.Exception
import Glue.IR (IR (..))

{- | Strict left-to-right fold over a list.

Curried signature: @foldl func initVal list@

Arguments:
  * @func@: The accumulator function applied at each step.
  * @initVal@: The initial value of the accumulator.
  * @list@: The target @List@ to fold over.

Lambda argument order: @(acc, x)@
  * @acc@: The current accumulator value (first parameter).
  * @x@: The current element of the list (second parameter).
-}
foldl :: IR Eval
foldl = NativeFunc foldlImpl

foldlImpl :: IR Eval -> Eval (IR Eval)
foldlImpl func = pure $ NativeFunc (foldlInit func)

foldlInit :: IR Eval -> IR Eval -> Eval (IR Eval)
foldlInit func initVal = pure $ NativeFunc (foldlOver func initVal)

foldlOver :: IR Eval -> IR Eval -> IR Eval -> Eval (IR Eval)
foldlOver func initVal list = case list of
    List xs -> do
        -- Strict left-to-right fold accumulating effects/values via foldM
        foldM (\acc x -> apply func [acc, x]) initVal xs
    _ -> throwError $ wrongArgumentType ["function", "any", "list"]

File: dart/glue/lib/src/lib/list/foldl.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Foldl function - strict left-to-right fold
Ir foldl = IrNativeFunc(foldlImpl);

Eval<Ir> foldlImpl(Ir funcIr) {
  return Eval.pure(IrNativeFunc(foldlInit(funcIr)));
}

Eval<Ir> Function(Ir) foldlInit(Ir funcIr) {
  return (Ir initValIr) {
    return Eval.pure(IrNativeFunc(foldlOver(funcIr, initValIr)));
  };
}

Eval<Ir> Function(Ir) foldlOver(Ir funcIr, Ir initValIr) {
  return (Ir listIr) {
    return switch (listIr) {
      IrList(elements: final elements) => _foldlLoop(
        funcIr,
        initValIr,
        elements.unlock,
      ),
      _ => throwError(wrongArgumentType(['function', 'any', 'list'])),
    };
  };
}

Eval<Ir> _foldlLoop(Ir funcIr, Ir acc, List<Ir> elements) {
  if (elements.isEmpty) {
    return Eval.pure(acc);
  }
  // Передаем аккумулятор и текущий элемент в функцию: f(acc, x)
  return apply(funcIr, [acc, elements.first]).bind((newAcc) {
    return _foldlLoop(funcIr, newAcc, elements.sublist(1));
  });
}
